File: src/chapters/linear-systems/content/code/gauss-complete.cpp
#include <vector>
#include <iostream>
#include <cmath>
#include <numeric>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Gaussian elimination with complete (row + column) pivoting.
Vec gauss_complete_pivot(Mat A, Vec b) {
    int n = b.size();
    vector<int> col(n); iota(col.begin(), col.end(), 0);
    for (int k = 0; k < n; ++k) {
        int pi = k, pj = k;
        for (int i = k; i < n; ++i)
            for (int j = k; j < n; ++j)
                if (fabs(A[i][j]) > fabs(A[pi][pj])) { pi = i; pj = j; }
        swap(A[k], A[pi]); swap(b[k], b[pi]);
        for (int i = 0; i < n; ++i) swap(A[i][k], A[i][pj]);
        swap(col[k], col[pj]);
        for (int i = k + 1; i < n; ++i) {
            double f = A[i][k] / A[k][k];
            for (int j = k; j < n; ++j) A[i][j] -= f * A[k][j];
            b[i] -= f * b[k];
        }
    }
    Vec y(n);
    for (int i = n - 1; i >= 0; --i) {
        double s = b[i];
        for (int j = i + 1; j < n; ++j) s -= A[i][j] * y[j];
        y[i] = s / A[i][i];
    }
    Vec x(n);
    for (int i = 0; i < n; ++i) x[col[i]] = y[i];   // undo column swaps
    return x;
}

int main() {
    Mat A = {{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}};
    Vec b = {8, -11, -3};
    Vec x = gauss_complete_pivot(A, b);
    cout << "x ="; for (double v : x) cout << " " << v; cout << "\n";
}

File: src/chapters/linear-systems/content/code/gauss-complete.py
import numpy as np


def gauss_complete_pivot(A, b):
    """Gaussian elimination with complete (row+column) pivoting.
    Returns x (column swaps are undone before returning)."""
    A = np.array(A, float)
    b = np.array(b, float)
    n = len(b)
    col = np.arange(n)                            # tracks column permutation
    for k in range(n):
        sub = np.abs(A[k:, k:])
        i, j = np.unravel_index(np.argmax(sub), sub.shape)
        i += k
        j += k
        A[[k, i]] = A[[i, k]]; b[[k, i]] = b[[i, k]]
        A[:, [k, j]] = A[:, [j, k]]; col[[k, j]] = col[[j, k]]
        for r in range(k + 1, n):
            f = A[r, k] / A[k, k]
            A[r, k:] -= f * A[k, k:]
            b[r] -= f * b[k]
    y = np.zeros(n)
    for i in range(n - 1, -1, -1):
        y[i] = (b[i] - A[i, i + 1:] @ y[i + 1:]) / A[i, i]
    x = np.zeros(n)
    x[col] = y                                    # undo column permutation
    return x


if __name__ == "__main__":
    A = [[2, 1, -1], [-3, -1, 2], [-2, 1, 2]]
    b = [8, -11, -3]
    print("x =", gauss_complete_pivot(A, b))
